sormas-backend/src/main/java/de/symeda/sormas/backend/document/DocumentFacadeEjb.java

```diff
@@ -0,0 +1,154 @@
+/*
+ * SORMAS® - Surveillance Outbreak Response Management & Analysis System
+ * Copyright © 2016-2020 Helmholtz-Zentrum für Infektionsforschung GmbH (HZI)
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <https://www.gnu.org/licenses/>.
+ */
+package de.symeda.sormas.backend.document;
+
+import java.io.IOException;
+import java.sql.Timestamp;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import javax.ejb.EJB;
+import javax.ejb.LocalBean;
+import javax.ejb.Stateless;
+import javax.enterprise.event.Event;
+import javax.inject.Inject;
+import javax.persistence.EntityExistsException;
+import javax.persistence.EntityManager;
+import javax.persistence.PersistenceContext;
+
+import de.symeda.sormas.api.document.DocumentDto;
+import de.symeda.sormas.api.document.DocumentFacade;
+import de.symeda.sormas.api.document.DocumentRelatedEntityType;
+import de.symeda.sormas.backend.user.UserFacadeEjb;
+import de.symeda.sormas.backend.user.UserService;
+import de.symeda.sormas.backend.util.DtoHelper;
+import de.symeda.sormas.backend.util.ModelConstants;
+
+@Stateless(name = "DocumentFacade")
+public class DocumentFacadeEjb implements DocumentFacade {
+
+	@PersistenceContext(unitName = ModelConstants.PERSISTENCE_UNIT_NAME)
+	private EntityManager em;
+
+	@EJB
+	private UserService userService;
+	@EJB
+	private DocumentService documentService;
+	@EJB
+	private DocumentStorageService documentStorageService;
+
+	@Inject
+	private Event<DocumentSaved> documentSavedEvent;
+
+	@Override
+	public DocumentDto getDocumentByUuid(String uuid) {
+		return toDto(documentService.getByUuid(uuid));
+	}
+
+	@Override
+	public DocumentDto saveDocument(DocumentDto dto, byte[] content) throws IOException {
+		Document existingDocument = dto.getUuid() == null ? null : documentService.getByUuid(dto.getUuid());
+		if (existingDocument != null) {
+			// TODO: add exception message
+			throw new EntityExistsException();
+		}
+
+		Document document = fromDto(dto);
+
+		document.setStorageReference(documentStorageService.save(document, content));
+
+		documentService.persist(document);
+		documentService.doFlush();
+
+		documentSavedEvent.fire(new DocumentSaved(document));
```

Comment:
The deletion of documents based on TransactionPhase.AFTER_FAILURE doesn't seem to be working.
I tried this by not full-filling one of the database constraints. The entity is correctly not saved in the database, but the stored document is not deleted as it should be.

Since you are doing "doFlush" here anyway, you should be able to directly catch any exception and trigger the deletion of the document from here.

Reply:
Done